### PR TITLE
RegexArrayShapeMatcher - fix capturing item-array-shapes for preg_match_all

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

File: src/Type/Php/RegexArrayShapeMatcher.php [hunk 1]
```diff
@@ -394,15 +394,19 @@ private function isGroupOptional(RegexCapturingGroup $captureGroup, TrinaryLogic
...
-			if (!$isTrailingOptional && $this->containsUnmatchedAsNull($flags, $matchesAll) && !$captureGroup->isOptional()) {
+			if (!$this->containsSetOrder($flags) && !$this->containsUnmatchedAsNull($flags, $matchesAll) && $captureGroup->isOptional()) {
+				$groupValueType = $this->getValueType(
+					TypeCombinator::union($captureGroup->getType(), new ConstantStringType('')),
+					$flags,
+					$matchesAll,
+				);
 				$groupValueType = TypeCombinator::removeNull($groupValueType);
+			} else {
+				$groupValueType = $this->getValueType($captureGroup->getType(), $flags, $matchesAll);
```
Comment (staabm) [inline src/Type/Php/RegexArrayShapeMatcher.php:405]: same as https://github.com/phpstan/phpstan-src/pull/3306/files#diff-fad531a4327d27f3f1fe6cf54631c80f2f2a4d15812f2ecafce861d6d7c0c67cR415-R423 but for `preg_match_all`

File: src/Type/Php/RegexArrayShapeMatcher.php [hunk 2]
```diff
@@ -474,6 +478,8 @@ private function getValueType(Type $baseType, int $flags, bool $matchesAll): Typ
 		$offsetType = IntegerRangeType::fromInterval(0, null);
 		if ($this->containsUnmatchedAsNull($flags, $matchesAll)) {
 			$valueType = TypeCombinator::addNull($valueType);
+		}
+		if ($this->containsUnmatchedAsNull($flags, $matchesAll) || $matchesAll) {
 			// unmatched groups return -1 as offset
 			$offsetType = IntegerRangeType::fromInterval(-1, null);
```
Comment (staabm) [inline src/Type/Php/RegexArrayShapeMatcher.php]: `preg_match_all` can yield `-1` in capturing offset, even without `PREG_UNMATCHED_AS_NULL`

https://3v4l.org/qU6Fa

Comment (staabm) [reply src/Type/Php/RegexArrayShapeMatcher.php]: hmm.. the same is true for `preg_match` - https://3v4l.org/Yra1j

**Comment by staabm** [comment]
phpstan-nette needs expectation adjustments

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**